File: sdks/typescript/src/webhooks/validation.ts
import { timingSafeEqual } from "node:crypto";

import { CregisClientError } from "../core/errors.js";
import { signProjectParameters } from "../signing/project.js";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function parseJsonObject(rawBody: string): Record<string, unknown> {
  if (typeof rawBody !== "string" || rawBody.trim() === "") {
    throw new CregisClientError("Callback body is required");
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(rawBody) as unknown;
  } catch (error) {
    throw new CregisClientError("Failed to parse callback JSON for signature verification", {
      cause: error,
    });
  }
  if (!isRecord(parsed)) {
    throw new CregisClientError("Callback body must be a JSON object");
  }
  return parsed;
}

function assertPositiveInteger(value: unknown, field: string): asserts value is number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value <= 0) {
    throw new CregisClientError(`Callback ${field} must be a positive integer`);
  }
}

export function verifyProjectWebhook(
  rawBody: string,
  apiKey: string,
): Record<string, unknown> {
  if (typeof apiKey !== "string" || apiKey.trim() === "") {
    throw new CregisClientError("API Key is required");
  }
  const parsed = parseJsonObject(rawBody);
  const incomingSign = parsed.sign;
  if (typeof incomingSign !== "string" || !/^[0-9a-f]{32}$/i.test(incomingSign)) {
    throw new CregisClientError("Callback signature must be a 32-character hexadecimal string");
  }
  assertPositiveInteger(parsed.pid, "pid");
  if (typeof parsed.nonce !== "string" || parsed.nonce.trim() === "") {
    throw new CregisClientError("Callback nonce must be a non-empty string");
  }
  assertPositiveInteger(parsed.timestamp, "timestamp");

  const unsigned = { ...parsed };
  delete unsigned.sign;
  const calculated = signProjectParameters(unsigned, apiKey);
  const incomingBuffer = Buffer.from(incomingSign.toLowerCase(), "ascii");
  const calculatedBuffer = Buffer.from(calculated, "ascii");
  if (!timingSafeEqual(calculatedBuffer, incomingBuffer)) {
    throw new CregisClientError("Callback signature verification failed");
  }
  return parsed;
}

File: sdks/python/src/cregis/webhooks/validation.py
"""Shared project webhook signature verification."""

from __future__ import annotations

import hmac
import json
import re
from typing import Any, Dict, Union

from cregis.errors import CregisClientError
from cregis.signing import sign_project_parameters

_SIGNATURE_PATTERN = re.compile(r"^[0-9a-f]{32}$", re.IGNORECASE)
RawBody = Union[str, bytes]


def verify_project_webhook(raw_body: RawBody, api_key: str) -> Dict[str, Any]:
    """Verify a Payment/WaaS callback and return its decoded JSON object."""

    if not isinstance(api_key, str) or not api_key.strip():
        raise CregisClientError("API Key is required")
    if not isinstance(raw_body, (str, bytes)) or not raw_body.strip():
        raise CregisClientError("Callback body is required")
    try:
        parsed = json.loads(raw_body)
    except (json.JSONDecodeError, UnicodeDecodeError, TypeError) as exc:
        raise CregisClientError("Failed to parse callback JSON for signature verification") from exc
    if not isinstance(parsed, dict):
        raise CregisClientError("Callback body must be a JSON object")

    incoming_sign = parsed.get("sign")
    if not isinstance(incoming_sign, str) or _SIGNATURE_PATTERN.fullmatch(incoming_sign) is None:
        raise CregisClientError("Callback signature must be a 32-character hexadecimal string")
    for field in ("pid", "timestamp"):
        value = parsed.get(field)
        if isinstance(value, bool) or not isinstance(value, int) or value <= 0:
            raise CregisClientError(f"Callback {field} must be a positive integer")
    nonce = parsed.get("nonce")
    if not isinstance(nonce, str) or not nonce.strip():
        raise CregisClientError("Callback nonce must be a non-empty string")

    unsigned = dict(parsed)
    del unsigned["sign"]
    calculated = sign_project_parameters(unsigned, api_key)
    if not hmac.compare_digest(calculated, incoming_sign.lower()):
        raise CregisClientError("Callback signature verification failed")
    return parsed
